### PR TITLE
Do not swallow exceptions in CLIEngine.getFormatter

### PR DESCRIPTION
**Pull request by gustavnikolaj**
Fix for issue #5977 

If a formatter is broken, the exception telling you that it was broken would be swallowed by the catch block, and the error reporting would tell you that it could not find the formatter.

This fix means that any exception, that is not a result of the formatter module not being found, will be re-thrown.

### CODE REVIEW

**Comment by gustavnikolaj** [comment]
The two failed builds on travis (node version [5](https://travis-ci.org/eslint/eslint/jobs/126083326) and [6](https://travis-ci.org/eslint/eslint/jobs/126083327)) both fail trying to install phantomjs when running npm install.

File: lib/cli-engine.js [hunk 1]
```diff
@@ -489,7 +489,10 @@ CLIEngine.getFormatter = function(format) {
         try {
             return require(formatterPath);
         } catch (ex) {
-            return null;
+            if (ex.code === "MODULE_NOT_FOUND" && ex.message.indexOf("'" + formatterPath + "'") !== -1) {
```
Comment (nzakas) [inline lib/cli-engine.js]: I'm not sure this is the right behavior. In all cases, it seems like we want to say "There was a problem loading 'formatterPath`" at the front of the message and then rethrow the error. Similar to: https://github.com/eslint/eslint/blob/master/lib/config/config-file.js#L146

File: tests/lib/cli-engine.js [hunk 2]
```diff
@@ -446,7 +446,7 @@ describe("CLIEngine", function() {
 
             var report = engine.executeOnFiles([getFixturePath("formatters")]);
 
-            assert.equal(report.results.length, 2);
+            assert.equal(report.results.length, 3);
```
Comment (nzakas) [inline tests/lib/cli-engine.js]: Why did this change?

Comment (gustavnikolaj) [inline tests/lib/cli-engine.js]: The test is running against all files in the formatters fixtures directory. I had to add another fixture (formatters/broken.js) which turns up in the generated report:

```js
console.log(report)
// { results: 
//    [ { filePath: '/tmp/eslint/fixtures/formatters/broken.js',
//        messages: [],
//        errorCount: 0,
//        warningCount: 0 },
//      { filePath: '/tmp/eslint/fixtures/formatters/simple.js',
//        messages: [],
//        errorCount: 0,
//        warningCount: 0 },
//      { filePath: '/tmp/eslint/fixtures/formatters/test/simple.js',
//        messages: [],
//        errorCount: 0,
//        warningCount: 0 } ],
//   errorCount: 0,
//   warningCount: 0 }
```

I probably should add the missing asserts for the 3rd item in the array to fix the rest of the test up.

File: lib/cli-engine.js (hunk 1 above)
Comment (gustavnikolaj) [inline lib/cli-engine.js]: That sounds like a good idea. The reason I didn't go there was that it is more of a breaking change, than just relaying some errors. I added a new commit on this branch where I make this change.

**Comment by gustavnikolaj** [comment]
I added a few new commits in response to @nzakas comments. I'll wait squashing until a decision is made so the changes and discussion are easier to follow. :-)

File: tests/lib/cli-engine.js [hunk 3]
```diff
@@ -1901,16 +1904,46 @@ describe("CLIEngine", function() {
...
+            assert.equal(error.message, "There was a problem loading formatter: " + formatterPath + "\nError: Cannot find module '" + formatterPath + "'");
         });
 
         it("should return null when a built-in formatter doesn't exist", function() {
             var engine = new CLIEngine(),
-                formatter = engine.getFormatter("special");
+                error = null;
 
-            assert.isNull(formatter);
+            try {
```
Comment (nzakas) [inline tests/lib/cli-engine.js]: You can use `assert.throws()` rather than using `try-catch`.

Comment (gustavnikolaj) [inline tests/lib/cli-engine.js]: Latest commit changes all instances of try/catch to `assert.throws()`. :+1:

**Comment by nzakas** [comment]
Looks like the tests aren't quite right, can you take a look?

**Comment by gustavnikolaj** [comment]
I had to make a minor update in lib/cli.js as errors are now thrown, and not just result in a null being returned.

All the tests passed now - sorry for not catching that the last time. Let me know if there's any other changes you'd like to see - if not I'll squash the commits, so it lives up to the pr requirements.

**Comment by nzakas** [comment]
Lgtm. GitHub will let me squash directly now, so you don't have to. :)

Thanks!

**Comment by gustavnikolaj** [comment]
Oh, that's nice. Thanks :-)

**Pull request merged**